go/adbc/driver/bigquery/connection.go

```diff
@@ -0,0 +1,730 @@
+		c.prefetchConcurrency = int(value)
+		return nil
+	default:
+		return c.ConnectionImplBase.SetOptionInt(key, value)
+	}
+}
+
+func (c *connectionImpl) newClient(ctx context.Context) error {
+	if c.catalog == "" {
+		return adbc.Error{
+			Code: adbc.StatusInvalidArgument,
+			Msg:  "ProjectID is empty",
+		}
+	}
+	switch c.authType {
+	case OptionValueAuthTypeJSONCredentialFile, OptionValueAuthTypeJSONCredentialString, OptionValueAuthTypeUserAuthentication:
+		var credentials option.ClientOption
+		if c.authType == OptionValueAuthTypeJSONCredentialFile {
+			credentials = option.WithCredentialsFile(c.credentials)
+		} else if c.authType == OptionValueAuthTypeJSONCredentialString {
+			credentials = option.WithCredentialsJSON([]byte(c.credentials))
+		} else {
+			if c.clientID == "" {
+				return adbc.Error{
+					Code: adbc.StatusInvalidArgument,
+					Msg:  fmt.Sprintf("The `%s` parameter is empty", OptionStringAuthClientID),
+				}
+			}
+			if c.clientSecret == "" {
+				return adbc.Error{
+					Code: adbc.StatusInvalidArgument,
+					Msg:  fmt.Sprintf("The `%s` parameter is empty", OptionStringAuthClientSecret),
+				}
+			}
+			if c.refreshToken == "" {
+				return adbc.Error{
+					Code: adbc.StatusInvalidArgument,
+					Msg:  fmt.Sprintf("The `%s` parameter is empty", OptionStringAuthRefreshToken),
+				}
+			}
+			credentials = option.WithTokenSource(c)
+		}
+
+		client, err := bigquery.NewClient(ctx, c.catalog, credentials)
+		if err != nil {
+			return err
+		}
+
+		err = client.EnableStorageReadClient(ctx, credentials)
+		if err != nil {
+			return err
+		}
+
+		c.client = client
+	default:
+		client, err := bigquery.NewClient(ctx, c.catalog)
+		if err != nil {
+			return err
+		}
+
+		err = client.EnableStorageReadClient(ctx)
+		if err != nil {
+			return err
+		}
+
+		c.client = client
+	}
+	return nil
+}
+
+var (
+	// Dataset:
+	//
+	// https://cloud.google.com/bigquery/docs/datasets#dataset-naming
+	//
+	// When you create a dataset in BigQuery, the dataset name must be unique for each project.
+	// The dataset name can contain the following:
+	//   - Up to 1,024 characters.
+	//   - Letters (uppercase or lowercase), numbers, and underscores.
+	// Dataset names are case-sensitive by default. mydataset and MyDataset can coexist in the same project,
+	// unless one of them has case-sensitivity turned off.
+	// Dataset names cannot contain spaces or special characters such as -, &, @, or %.
+	datasetRegex = regexp.MustCompile("^[a-zA-Z0-9_-]")
+
+	precisionScaleRegex = regexp.MustCompile(`^(?:BIG)?NUMERIC\((?P<precision>\d+)(?:,(?P<scale>\d+))?\)$`)
+	simpleDataType      = map[string]arrow.DataType{
+		"BOOL":      arrow.FixedWidthTypes.Boolean,
+		"BOOLEAN":   arrow.FixedWidthTypes.Boolean,
+		"FLOAT":     arrow.PrimitiveTypes.Float64,
+		"FLOAT64":   arrow.PrimitiveTypes.Float64,
+		"BYTES":     arrow.BinaryTypes.Binary,
+		"STRING":    arrow.BinaryTypes.String,
+		"GEOGRAPHY": arrow.BinaryTypes.String,
```

Comment:
not needed for this PR, but as a `TODO` we should consider using GeoArrow potentially for this